{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import sys\n",
    "sys.path.append('../src')\n",
    "\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set_style('white')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table S1 - Sample overview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>t2onc_type</th>\n",
       "      <th>dnaseq_run</th>\n",
       "      <th>rnaseq_run</th>\n",
       "      <th>rnaseq_id</th>\n",
       "      <th>mouse</th>\n",
       "      <th>pathology_type</th>\n",
       "      <th>include_shearsplink</th>\n",
       "      <th>metastasis_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11KOU012-R3</td>\n",
       "      <td>chr15-donor</td>\n",
       "      <td>SBecad</td>\n",
       "      <td>run120</td>\n",
       "      <td>1566_6_11KOU012-R3</td>\n",
       "      <td>11KOU012</td>\n",
       "      <td>Spindle cell tumor, ILC</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11KOU012-R5</td>\n",
       "      <td>chr15-donor</td>\n",
       "      <td>SBecad</td>\n",
       "      <td>run120</td>\n",
       "      <td>1566_7_11KOU012-R5</td>\n",
       "      <td>11KOU012</td>\n",
       "      <td>ILC</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11KOU015</td>\n",
       "      <td>chr1-donor</td>\n",
       "      <td>SBecad</td>\n",
       "      <td>run120</td>\n",
       "      <td>1566_8_11KOU015</td>\n",
       "      <td>11KOU015</td>\n",
       "      <td>Spindle cell tumor</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11KOU018</td>\n",
       "      <td>chr1-donor</td>\n",
       "      <td>SBecad</td>\n",
       "      <td>run120</td>\n",
       "      <td>1566_9_11KOU018</td>\n",
       "      <td>11KOU018</td>\n",
       "      <td>Squamous, spindle cell, ILC</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11KOU023</td>\n",
       "      <td>chr15-donor</td>\n",
       "      <td>SBecad</td>\n",
       "      <td>run120</td>\n",
       "      <td>1566_10_11KOU023</td>\n",
       "      <td>11KOU023</td>\n",
       "      <td>ILC</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample   t2onc_type dnaseq_run rnaseq_run           rnaseq_id  \\\n",
       "0  11KOU012-R3  chr15-donor     SBecad     run120  1566_6_11KOU012-R3   \n",
       "1  11KOU012-R5  chr15-donor     SBecad     run120  1566_7_11KOU012-R5   \n",
       "2     11KOU015   chr1-donor     SBecad     run120     1566_8_11KOU015   \n",
       "3     11KOU018   chr1-donor     SBecad     run120     1566_9_11KOU018   \n",
       "4     11KOU023  chr15-donor     SBecad     run120    1566_10_11KOU023   \n",
       "\n",
       "      mouse               pathology_type include_shearsplink metastasis_site  \n",
       "0  11KOU012      Spindle cell tumor, ILC                True             NaN  \n",
       "1  11KOU012                          ILC                True             NaN  \n",
       "2  11KOU015          Spindle cell tumor                False             NaN  \n",
       "3  11KOU018  Squamous, spindle cell, ILC                True             NaN  \n",
       "4  11KOU023                          ILC                True             NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read sample overview.\n",
    "samples = pd.read_csv('../data/raw/sb/samples.txt', sep='\\t')\n",
    "\n",
    "# Summarize metastases per mouse.\n",
    "metastases = pd.read_csv('../data/raw/sb/metastases.txt', sep='\\t')\n",
    "metastases = metastases.ix[metastases['mouse'].isin(samples['mouse'])]\n",
    "\n",
    "# Merge metastasis data with samples.\n",
    "met_sites = (metastases.groupby('mouse')\n",
    "                       .agg({'metastasis_site': lambda s: ', '.join(s)})\n",
    "                       .reset_index())\n",
    "\n",
    "samples = pd.merge(samples, met_sites, on='mouse', how='left')\n",
    "samples.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "! mkdir -p ../reports/supplemental/tables/\n",
    "samples.to_excel('../reports/supplemental/tables/table_s1_samples.xlsx', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table S2 - Candidate overview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>n_samples</th>\n",
       "      <th>mean_clonality</th>\n",
       "      <th>sense_fraction</th>\n",
       "      <th>sense_fraction_weighted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Arfip1</td>\n",
       "      <td>7</td>\n",
       "      <td>0.809442</td>\n",
       "      <td>0.636364</td>\n",
       "      <td>0.747715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Arid1a</td>\n",
       "      <td>14</td>\n",
       "      <td>0.399860</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>0.181567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Asxl2</td>\n",
       "      <td>5</td>\n",
       "      <td>0.520660</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.373262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bach2</td>\n",
       "      <td>3</td>\n",
       "      <td>0.392943</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cblb</td>\n",
       "      <td>3</td>\n",
       "      <td>0.443457</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name  n_samples  mean_clonality  sense_fraction  \\\n",
       "0    Arfip1          7        0.809442        0.636364   \n",
       "1    Arid1a         14        0.399860        0.266667   \n",
       "2     Asxl2          5        0.520660        0.500000   \n",
       "3     Bach2          3        0.392943        1.000000   \n",
       "4      Cblb          3        0.443457        0.000000   \n",
       "\n",
       "   sense_fraction_weighted  \n",
       "0                 0.747715  \n",
       "1                 0.181567  \n",
       "2                 0.373262  \n",
       "3                 1.000000  \n",
       "4                 0.000000  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from nbsupport import insertions as nb_ins\n",
    "\n",
    "# Read insertions.\n",
    "insertions = (pd.read_csv('../data/processed/sb/shear_splink/'\n",
    "                          'all/insertions.cis.rbm.txt', sep='\\t')\n",
    "                .pipe(nb_ins.annotate_with_clonality))\n",
    "\n",
    "# Summarize various statistics (including donor bias) for genes\n",
    "# that were identified in the overall CIS analysis.\n",
    "ranked_summary = nb_ins.gene_statistics(insertions)\n",
    "\n",
    "ranked_summary.index.name = 'gene_name'\n",
    "ranked_summary = ranked_summary.reset_index()\n",
    "\n",
    "ranked_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mt-Tp</td>\n",
       "      <td>MT</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mt-Tt</td>\n",
       "      <td>MT</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mt-Cytb</td>\n",
       "      <td>MT</td>\n",
       "      <td>14145</td>\n",
       "      <td>15288</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mt-Te</td>\n",
       "      <td>MT</td>\n",
       "      <td>14071</td>\n",
       "      <td>14139</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mt-Nd6</td>\n",
       "      <td>MT</td>\n",
       "      <td>13552</td>\n",
       "      <td>14070</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name chromosome_name  start_position  end_position  strand\n",
       "0     mt-Tp              MT           15356         15422      -1\n",
       "1     mt-Tt              MT           15289         15355       1\n",
       "2   mt-Cytb              MT           14145         15288       1\n",
       "3     mt-Te              MT           14071         14139      -1\n",
       "4    mt-Nd6              MT           13552         14070      -1"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pybiomart\n",
    "\n",
    "# Annotate location using biomart.\n",
    "bm_dataset = pybiomart.Dataset(name='mmusculus_gene_ensembl',\n",
    "                               host='http://www.ensembl.org')\n",
    "\n",
    "bm_annotation = bm_dataset.query(\n",
    "    attributes=['external_gene_name', 'chromosome_name',\n",
    "                'start_position', 'end_position', 'strand'],\n",
    "    use_attr_names=True)\n",
    "\n",
    "bm_annotation = bm_annotation.rename(\n",
    "    columns={'external_gene_name': 'gene_name'})\n",
    "\n",
    "bm_annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>n_samples_chr1_donor</th>\n",
       "      <th>n_samples_chr15_donor</th>\n",
       "      <th>donor_bias_pval</th>\n",
       "      <th>donor_bias_qval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Trp53bp2</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.042451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Runx1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0.008175</td>\n",
       "      <td>0.122619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ppp1r12a</td>\n",
       "      <td>11</td>\n",
       "      <td>30</td>\n",
       "      <td>0.022114</td>\n",
       "      <td>0.201767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rgag1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.026902</td>\n",
       "      <td>0.201767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Trp53</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.039225</td>\n",
       "      <td>0.235351</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name  n_samples_chr1_donor  n_samples_chr15_donor  donor_bias_pval  \\\n",
       "0  Trp53bp2                    12                      3         0.001415   \n",
       "1     Runx1                     7                      1         0.008175   \n",
       "2  Ppp1r12a                    11                     30         0.022114   \n",
       "3     Rgag1                     4                      0         0.026902   \n",
       "4     Trp53                     0                      7         0.039225   \n",
       "\n",
       "   donor_bias_qval  \n",
       "0         0.042451  \n",
       "1         0.122619  \n",
       "2         0.201767  \n",
       "3         0.201767  \n",
       "4         0.235351  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from nbsupport.morphology import test_strain_bias\n",
    "\n",
    "strain_map = dict(zip(samples['sample'], samples['t2onc_type']))\n",
    "insertions['t2onc_type'] = insertions['sample'].map(strain_map)\n",
    "\n",
    "gene_bias = (\n",
    "    test_strain_bias(\n",
    "        insertions, value='gene_name', incl_neg=False,\n",
    "        samples=samples.query('include_shearsplink == True'))\n",
    "    .sort_values('p_value')\n",
    "    .rename(columns={\n",
    "            'pos_chr1-donor': 'n_samples_chr1_donor',\n",
    "            'pos_chr15-donor': 'n_samples_chr15_donor',\n",
    "            'p_value': 'donor_bias_pval',\n",
    "            'q_value': 'donor_bias_qval'\n",
    "        })\n",
    "    .reset_index())\n",
    "\n",
    "gene_bias.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>cis_chr15_donor</th>\n",
       "      <th>cis_chr1_donor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Arfip1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Arid1a</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Asxl2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bach2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cblb</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name cis_chr15_donor cis_chr1_donor\n",
       "0    Arfip1            True          False\n",
       "1    Arid1a            True           True\n",
       "2     Asxl2           False          False\n",
       "3     Bach2           False          False\n",
       "4      Cblb           False          False"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "insertions_chr15 = pd.read_csv('../data/processed/sb/shear_splink/1868/'\n",
    "                              'insertions.cis.rbm.txt', sep='\\t')\n",
    "\n",
    "insertions_chr1 = pd.read_csv('../data/processed/sb/shear_splink/1869/'\n",
    "                              'insertions.cis.rbm.txt', sep='\\t')\n",
    "\n",
    "cis_bias = pd.DataFrame({'gene_name': ranked_summary['gene_name']})\n",
    "cis_bias['cis_chr15_donor'] = cis_bias['gene_name'].isin(insertions_chr15['gene_name'])\n",
    "cis_bias['cis_chr1_donor'] = cis_bias['gene_name'].isin(insertions_chr1['gene_name'])\n",
    "\n",
    "cis_bias.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>strand</th>\n",
       "      <th>n_samples</th>\n",
       "      <th>mean_clonality</th>\n",
       "      <th>sense_fraction</th>\n",
       "      <th>sense_fraction_weighted</th>\n",
       "      <th>n_samples_chr1_donor</th>\n",
       "      <th>n_samples_chr15_donor</th>\n",
       "      <th>donor_bias_pval</th>\n",
       "      <th>donor_bias_qval</th>\n",
       "      <th>cis_chr15_donor</th>\n",
       "      <th>cis_chr1_donor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Fgfr2</td>\n",
       "      <td>7</td>\n",
       "      <td>130162451</td>\n",
       "      <td>133123350</td>\n",
       "      <td>-1</td>\n",
       "      <td>56</td>\n",
       "      <td>0.761385</td>\n",
       "      <td>0.513889</td>\n",
       "      <td>0.492919</td>\n",
       "      <td>26</td>\n",
       "      <td>30</td>\n",
       "      <td>0.305059</td>\n",
       "      <td>0.787224</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Trps1</td>\n",
       "      <td>15</td>\n",
       "      <td>50654752</td>\n",
       "      <td>50890463</td>\n",
       "      <td>-1</td>\n",
       "      <td>51</td>\n",
       "      <td>0.618730</td>\n",
       "      <td>0.103704</td>\n",
       "      <td>0.055212</td>\n",
       "      <td>19</td>\n",
       "      <td>32</td>\n",
       "      <td>0.419853</td>\n",
       "      <td>0.787224</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Ppp1r12a</td>\n",
       "      <td>10</td>\n",
       "      <td>108162400</td>\n",
       "      <td>108277575</td>\n",
       "      <td>1</td>\n",
       "      <td>41</td>\n",
       "      <td>0.552525</td>\n",
       "      <td>0.021277</td>\n",
       "      <td>0.003714</td>\n",
       "      <td>11</td>\n",
       "      <td>30</td>\n",
       "      <td>0.022114</td>\n",
       "      <td>0.201767</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Myh9</td>\n",
       "      <td>15</td>\n",
       "      <td>77760587</td>\n",
       "      <td>77842175</td>\n",
       "      <td>-1</td>\n",
       "      <td>26</td>\n",
       "      <td>0.794418</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>0.150643</td>\n",
       "      <td>8</td>\n",
       "      <td>18</td>\n",
       "      <td>0.249404</td>\n",
       "      <td>0.787224</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Trp53bp2</td>\n",
       "      <td>1</td>\n",
       "      <td>182409172</td>\n",
       "      <td>182462432</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>0.755975</td>\n",
       "      <td>0.842105</td>\n",
       "      <td>0.868771</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.042451</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_name chromosome_name  start_position  end_position  strand  n_samples  \\\n",
       "11     Fgfr2               7       130162451     133123350      -1         56   \n",
       "10     Trps1              15        50654752      50890463      -1         51   \n",
       "24  Ppp1r12a              10       108162400     108277575       1         41   \n",
       "7       Myh9              15        77760587      77842175      -1         26   \n",
       "22  Trp53bp2               1       182409172     182462432       1         15   \n",
       "\n",
       "    mean_clonality  sense_fraction  sense_fraction_weighted  \\\n",
       "11        0.761385        0.513889                 0.492919   \n",
       "10        0.618730        0.103704                 0.055212   \n",
       "24        0.552525        0.021277                 0.003714   \n",
       "7         0.794418        0.108108                 0.150643   \n",
       "22        0.755975        0.842105                 0.868771   \n",
       "\n",
       "    n_samples_chr1_donor  n_samples_chr15_donor  donor_bias_pval  \\\n",
       "11                    26                     30         0.305059   \n",
       "10                    19                     32         0.419853   \n",
       "24                    11                     30         0.022114   \n",
       "7                      8                     18         0.249404   \n",
       "22                    12                      3         0.001415   \n",
       "\n",
       "    donor_bias_qval cis_chr15_donor cis_chr1_donor  \n",
       "11         0.787224            True           True  \n",
       "10         0.787224            True           True  \n",
       "24         0.201767            True           True  \n",
       "7          0.787224            True           True  \n",
       "22         0.042451            True           True  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge everything.\n",
    "candidate_overview = (\n",
    "    bm_annotation\n",
    "    .pipe(pd.merge, right=ranked_summary, on='gene_name', how='right')\n",
    "    .pipe(pd.merge, right=gene_bias, on='gene_name', how='left')\n",
    "    .pipe(pd.merge, right=cis_bias, on='gene_name', how='left')\n",
    "    .sort_values('n_samples', ascending=False))\n",
    "\n",
    "# Re-order columns for legibility.\n",
    "#candidate_overview = candidate_overview[\n",
    "#    ['gene_name', 'chromosome_name', 'start_position', 'end_position',\n",
    "#     'strand', 'n_samples', 'n_samples_chr15_donor', 'n_samples_chr1_donor',\n",
    "#     'donor_bias_pval', 'donor_bias_qval', 'cis_chr15_donor', 'cis_chr1_donor',\n",
    "#     'mean_clonality', 'sense_fraction_weighted']]\n",
    "\n",
    "# Sort by frequency/clonality.\n",
    "candidate_overview.sort_values(\n",
    "    ['n_samples', 'mean_clonality'],\n",
    "    ascending=False, inplace=True)\n",
    "\n",
    "candidate_overview.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "candidate_overview.to_excel('../reports/supplemental/tables/table_s2_candidate_overview.xlsx', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>strand</th>\n",
       "      <th>n_samples</th>\n",
       "      <th>mean_clonality</th>\n",
       "      <th>sense_fraction</th>\n",
       "      <th>sense_fraction_weighted</th>\n",
       "      <th>n_samples_chr1_donor</th>\n",
       "      <th>n_samples_chr15_donor</th>\n",
       "      <th>donor_bias_pval</th>\n",
       "      <th>donor_bias_qval</th>\n",
       "      <th>cis_chr15_donor</th>\n",
       "      <th>cis_chr1_donor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Trp53bp2</td>\n",
       "      <td>1</td>\n",
       "      <td>182409172</td>\n",
       "      <td>182462432</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>0.755975</td>\n",
       "      <td>0.842105</td>\n",
       "      <td>0.868771</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.042451</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Ppp1r12b</td>\n",
       "      <td>1</td>\n",
       "      <td>134754658</td>\n",
       "      <td>134955942</td>\n",
       "      <td>-1</td>\n",
       "      <td>7</td>\n",
       "      <td>0.442771</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_name chromosome_name  start_position  end_position  strand  n_samples  \\\n",
       "22  Trp53bp2               1       182409172     182462432       1         15   \n",
       "16  Ppp1r12b               1       134754658     134955942      -1          7   \n",
       "\n",
       "    mean_clonality  sense_fraction  sense_fraction_weighted  \\\n",
       "22        0.755975        0.842105                 0.868771   \n",
       "16        0.442771        0.000000                 0.000000   \n",
       "\n",
       "    n_samples_chr1_donor  n_samples_chr15_donor  donor_bias_pval  \\\n",
       "22                    12                      3         0.001415   \n",
       "16                     3                      4         1.000000   \n",
       "\n",
       "    donor_bias_qval cis_chr15_donor cis_chr1_donor  \n",
       "22         0.042451            True           True  \n",
       "16         1.000000            True           True  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome 15\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>strand</th>\n",
       "      <th>n_samples</th>\n",
       "      <th>mean_clonality</th>\n",
       "      <th>sense_fraction</th>\n",
       "      <th>sense_fraction_weighted</th>\n",
       "      <th>n_samples_chr1_donor</th>\n",
       "      <th>n_samples_chr15_donor</th>\n",
       "      <th>donor_bias_pval</th>\n",
       "      <th>donor_bias_qval</th>\n",
       "      <th>cis_chr15_donor</th>\n",
       "      <th>cis_chr1_donor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Trps1</td>\n",
       "      <td>15</td>\n",
       "      <td>50654752</td>\n",
       "      <td>50890463</td>\n",
       "      <td>-1</td>\n",
       "      <td>51</td>\n",
       "      <td>0.618730</td>\n",
       "      <td>0.103704</td>\n",
       "      <td>0.055212</td>\n",
       "      <td>19</td>\n",
       "      <td>32</td>\n",
       "      <td>0.419853</td>\n",
       "      <td>0.787224</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Myh9</td>\n",
       "      <td>15</td>\n",
       "      <td>77760587</td>\n",
       "      <td>77842175</td>\n",
       "      <td>-1</td>\n",
       "      <td>26</td>\n",
       "      <td>0.794418</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>0.150643</td>\n",
       "      <td>8</td>\n",
       "      <td>18</td>\n",
       "      <td>0.249404</td>\n",
       "      <td>0.787224</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_name chromosome_name  start_position  end_position  strand  n_samples  \\\n",
       "10     Trps1              15        50654752      50890463      -1         51   \n",
       "7       Myh9              15        77760587      77842175      -1         26   \n",
       "\n",
       "    mean_clonality  sense_fraction  sense_fraction_weighted  \\\n",
       "10        0.618730        0.103704                 0.055212   \n",
       "7         0.794418        0.108108                 0.150643   \n",
       "\n",
       "    n_samples_chr1_donor  n_samples_chr15_donor  donor_bias_pval  \\\n",
       "10                    19                     32         0.419853   \n",
       "7                      8                     18         0.249404   \n",
       "\n",
       "    donor_bias_qval cis_chr15_donor cis_chr1_donor  \n",
       "10         0.787224            True           True  \n",
       "7          0.787224            True           True  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display\n",
    "\n",
    "# Print donor chromosome genes.\n",
    "for chrom in ['1', '15']:\n",
    "    print('Chromosome {}'.format(chrom))\n",
    "    display(candidate_overview.query('chromosome_name == {!r}'.format(chrom)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>n_samples_chr1_donor</th>\n",
       "      <th>n_samples_chr15_donor</th>\n",
       "      <th>donor_bias_pval</th>\n",
       "      <th>donor_bias_qval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Trp53bp2</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>0.001415</td>\n",
       "      <td>0.042451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Runx1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0.008175</td>\n",
       "      <td>0.122619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ppp1r12a</td>\n",
       "      <td>11</td>\n",
       "      <td>30</td>\n",
       "      <td>0.022114</td>\n",
       "      <td>0.201767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rgag1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.026902</td>\n",
       "      <td>0.201767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Trp53</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.039225</td>\n",
       "      <td>0.235351</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name  n_samples_chr1_donor  n_samples_chr15_donor  donor_bias_pval  \\\n",
       "0  Trp53bp2                    12                      3         0.001415   \n",
       "1     Runx1                     7                      1         0.008175   \n",
       "2  Ppp1r12a                    11                     30         0.022114   \n",
       "3     Rgag1                     4                      0         0.026902   \n",
       "4     Trp53                     0                      7         0.039225   \n",
       "\n",
       "   donor_bias_qval  \n",
       "0         0.042451  \n",
       "1         0.122619  \n",
       "2         0.201767  \n",
       "3         0.201767  \n",
       "4         0.235351  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_bias.query('donor_bias_pval < 0.05')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table S5 - Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sb_counts = pd.read_csv('../data/processed/sb/rnaseq/gene_counts.txt', sep='\\t', index_col=0)\n",
    "kb1p_counts = pd.read_csv('../data/processed/kb1p/gene_counts.txt', sep='\\t', index_col=0)\n",
    "pten_counts = pd.read_csv('../data/processed/pten/gene_counts.txt', sep='\\t', index_col=0)\n",
    "\n",
    "with pd.ExcelWriter('../reports/supplemental/tables/table_s5_expression.xlsx') as writer:\n",
    "    sb_counts.to_excel(writer, sheet_name='SB samples')\n",
    "    kb1p_counts.to_excel(writer, sheet_name='KB1P samples')\n",
    "    pten_counts.to_excel(writer, sheet_name='EcadPten samples')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table S6 - Insertions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "insertions = pd.read_csv('../data/processed/sb/shear_splink/all/insertions.txt', sep='\\t')\n",
    "insertions_annotated = pd.read_csv('../data/processed/sb/shear_splink/all/insertions.cis.rbm.txt', sep='\\t')\n",
    "\n",
    "cis_sites = pd.read_csv('../data/processed/sb/shear_splink/all/insertions.cis.sites.txt', sep='\\t')\n",
    "cis_insertions = pd.read_csv('../data/processed/sb/shear_splink/all/insertions.cis.txt', sep='\\t')\n",
    "\n",
    "cis_mapping = cis_insertions[['id', 'cis_id']]\n",
    "\n",
    "with pd.ExcelWriter('../reports/supplemental/tables/table_s6_insertions.xlsx') as writer:\n",
    "    insertions.to_excel(writer, sheet_name='insertions', index=False)\n",
    "    cis_sites.to_excel(writer, sheet_name='cis_sites', index=False)\n",
    "    cis_mapping.to_excel(writer, sheet_name='cis_mapping', index=False)\n",
    "    insertions_annotated.to_excel(writer, sheet_name='insertions_annotated', index=False)"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
